### PR TITLE
Test/calculation

### PR DESCRIPTION
**Pull request by Hsuwaddyhtat22**
unit tests with React Testing Library.
unit tests with snapshots with Jest.
Set up React Testing Library.
Followed the instructions in the official documentation.
Created unit tests for all React components:
Used Jest snapshots to test the components.
Used React Testing Library to simulate user interaction.

### CODE REVIEW

**Review by tufoinnkuo10 — CHANGES_REQUESTED** [review]
Hi @Hsuwaddyhtat22 

While you made a great effort in this project, unfortunately, I cannot proceed to review your code.

## Invalid Code Review Request

You have submitted a project, that Does not have contributions from your coding partner. please note that this was a pair programming activity, kindly collaborate with your partner, make commits together, and submit for review. 😄 

Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

**Comment by Hsuwaddyhtat22** [comment]
Andrea Funda (Microverse)

Sep 22, 2023, 16:15 GMT+2

Hi Wendy,
 
This late in the week this is no longer possible. All other teams are nearing the end of their project work for the week.
 
In the future, I'd like to advise you to please write to us when you first begin to experience collaborative issues with a partner. It's very difficult for us to support you and your partner when we're made aware of any issues so late in the week.

It's essential to discuss strategies and agree on working timelines for collaborative projects within the work week with your partner and, if you find that your partnership is not going smoothly, it's best to flag to our Support team as soon as it happens, especially at the beginning of the week (Monday or Tuesday) so that we may assist you with feedback strategies and in developing working agreements. 

This gives everyone a fair try at making things work and then, if there's absolutely no other option, it allows us to unpair you from a partner that we are all certain you cannot work alongside.
 
I'm also including a few articles that may be of interest:
/``

**Comment by Hsuwaddyhtat22** [comment]
Screenshot_2023-09-22-22-46-43-060_com google android gm

File: src/__test__/calculate.test.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  expect(displays.length).toBeGreaterThanOrEqual(1);
+});
+
+test('clicking AC button resets the display to "0"', () => {
+  render(<Calculator />);
+  fireEvent.click(screen.getByText('AC'));
+  // Check if at least one display resets to "0"
+  const displays = screen.getAllByText('0');
+  expect(displays.length).toBeGreaterThanOrEqual(1);
+});
```
Comment (FrederickMih) [inline src/__test__/calculate.test.js]: - [ ] Hi @Hsuwaddyhtat22, the string in the **test** block should be something that should actually happen with your tests. I think the matcher **toBeGreaterThanOrEqual** doesn't assert the hypothesis you made. I think it was preferable to use the `describe` block which wrap the `test` or `it` blocks and it is the best way to group our tests. Clicking on "AC", we expect the output to be zero :+1:

File: src/__test__/calculate.test.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+test('clicking number buttons updates the display', () => {
+  render(<Calculator />);
+  // Click on the number buttons '1' and '2'
+  fireEvent.click(screen.getByText('1'));
+  fireEvent.click(screen.getByText('2'));
+  // Check if at least one display shows '12'
+  const displays = screen.getAllByText('12');
+  expect(displays.length).toBeGreaterThanOrEqual(1);
+});
```
Comment (FrederickMih) [inline src/__test__/calculate.test.js]: - [ ] Once more let's kindly make sure that all the functionalities and scenarios of calculate.js module should be covered using jest. For this specific, you'll kindly make sure the following scenarios are covered

- 1

```
if (isNumber(buttonName)) {
    if (obj.operation === null) {
      if (obj.next === null) {
        return { ...obj, next: buttonName };
      } if (obj.next === '0') {
        return { ...obj, next: buttonName };
      }
      return { ...obj, next: obj.next + buttonName };
    }
    if (obj.next === null) {
      return { ...obj, next: buttonName };
    }
    return { ...obj, next: obj.next + buttonName };
  }
```

- 2

```
  if (buttonName === '.') {
    if (obj.next === null) {
      return { ...obj, next: '0.' };
    }
    if (!obj.next.includes('.')) {
      return { ...obj, next: `${obj.next}.` };
    }
    return { ...obj };
  }
```

- 3

```
if (buttonName === '=') {
    if (obj.next !== null && obj.operation !== null) {
      try {
        const total = operate(obj.total, obj.next, obj.operation);
        return {
          total: total.toString(),
          next: null,
          operation: null,
        };
      } catch (error) {
        return {
          total: 'Error',
          next: null,
          operation: null,
        };
      }
    }
    return { ...obj };
  }
```

- 4

```
  if (buttonName === '+/-') {
    if (obj.next !== null) {
      return { ...obj, next: (-parseFloat(obj.next)).toString() };
    }
    if (obj.total !== null) {
      return { ...obj, total: (-parseFloat(obj.total)).toString() };
    }
    return { ...obj };
  }
```

- 5

```
  if (['+', '-', '*', '/', '%'].includes(buttonName)) {
    if (obj.operation !== null) {
      try {
        const total = operate(obj.total, obj.next, obj.operation);
        return {
          total: total.toString(),
          next: null,
          operation: buttonName,
        };
      } catch (error) {
        return {
          total: 'Error',
          next: null,
          operation: null,
        };
      }
    }
```

- lastly

```
if (buttonName === '/') {
    if (obj.next === '0') {
      return {
        total: 'Error',
        next: null,
        operation: null,
      };
    }
  }
```

**Review by FrederickMih — CHANGES_REQUESTED** [review]
## Hi :raising_hand: :raising_hand: @Hsuwaddyhtat22, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!





## Highlights :green_circle: 

- operate.js module have been testing as expected :+1: 


## Required Changes ♻️

- [ ] Once more let's kindly make sure there's the usage of Jest snapshots for testing all the components. Right now your tests shows zero(0) snapshot.

_Check the comments under the review._

## Optional suggestions
**N/A**
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/__test__/calculate.test.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  fireEvent.click(screen.getByText('AC'));
+  expect(container).toMatchSnapshot();
+});
+
+test('clicking number buttons updates the display', () => {
+  const { container } = render(<Calculator />);
+  fireEvent.click(screen.getByText('1'));
+  fireEvent.click(screen.getByText('2'));
+  expect(container).toMatchSnapshot();
+});
```
Comment (danmainah) [inline src/__test__/calculate.test.js]: - [ ] @Hsuwaddyhtat22 as previously requested kindly ensure your tests cover all functionalities and scenarios of the calculate.js module, they should be 6 in total.

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @Hsuwaddyhtat22 , 

Great work on making the changes requested by a previous reviewer 👏  👏 👏
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
✔️  Implemented the change requested to ensure the use of  Jest snapshots when testing components.

 # Required Changes ♻️

- [ ] @Hsuwaddyhtat22 Kindly ensure all tests are passing after implementing the changes requested

_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by julie-ify — APPROVED** [review]
Hi @Hsuwaddyhtat22 and @tufoinnkuo10,

## Highlights!
- Good code organization and tests ✔️
- Descriptive README ✔️ 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# Status: Approved ✔️ 

Cheers and Happy coding!👏👏👏

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Feel free to leave any questions or comments in the PR thread if something is not 100% clear.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**